### PR TITLE
Add migration to handle webpush subscriptions on extension purge.

### LINKED ISSUE #5: Add migration for phpBB 4.0
For phpBB 4.0 boards we need to:

1) Disable this extension ?
2) Migrate it's table data over to phpBB 4.0's web push table data ?

### PR DESCRIPTION
**Pull request by rxu**
Closes #5.

### CODE REVIEW

**Comment by rxu** [comment]
TODO: add moving extension tables data over corresponding phpBB 4.0 tables etc.

**Comment by rxu** [comment]
This looks weird though so there should be some instruction on how to use this to avoid board mess and/or crash.
F.e. after upgrading board to phpBB 4.0 the extension should be purged 1st while the board is still disabled so that users didn't create any webpush subscriptions yet.

File: migrations/handle_subscriptions.php [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+			'DELETE FROM ' . $user_notifications_table . "
+				WHERE method = '" . $this->db->sql_escape('phpbb.wpn.notification.method.webpush') . "'";
+
+		$this->db->sql_query($sql);
+	}
+
+	public function copy_subscription_tables()
+	{
+		$core_notification_push_table = $this->table_prefix . 'notification_push';
```
Comment (iMattPro) [inline migrations/handle_subscriptions.php:53]: I feel like we should make sure the core tables exist for sure before doing this. Either checking here or in the IF statement in. the revert_data method above?

File: migrations/handle_subscriptions.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+ *
+ */
+
+namespace phpbb\webpushnotifications\migrations;
+
+use phpbb\db\migration\container_aware_migration;
+
+class handle_subscriptions extends container_aware_migration
+{
+	public function revert_data(): array
```
Comment (iMattPro) [inline migrations/handle_subscriptions.php:17]: We should probably have an effectively installed method too, so we make sure this only runs if our extension is installed (so in other words this won't run during an installation, only during a delete/purge). 

Maybe:
```php
public function effectively_installed()
{
	return !$this->db_tools->sql_table_exists($this->table_prefix . 'wpn_notification_push');
}
```

File: migrations/handle_subscriptions.php (hunk 1 above)
Comment (iMattPro) [reply migrations/handle_subscriptions.php:53]: Also should we copy over the API keys and enable/disabled configs too?

Comment (rxu) [reply migrations/handle_subscriptions.php:53]: > Also should we copy over the API keys and enable/disabled configs too?

Oh, should we then rename current ext config parameters to something `wpn`-prefixed?

**Comment by iMattPro** [comment]
Oh yeah. We missed that I guess. We should prefix everything that could possibly conflict with the 4.0 implantation

**Comment by rxu** [comment]
We probably need to unit-test this migration somehow, f.e. create fixtures with tables and configs including ones from 4.0 PR, run the migration and assert exported values.

**Review by iMattPro — APPROVED** [review]

File: migrations/handle_subscriptions.php [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+			['if', [
+				($this->db_tools->sql_table_exists($this->table_prefix . 'notification_push')),
+				['custom', [[$this, 'copy_subscription_tables']]],
+			]],
+			['if', [
+				(isset($this->config['webpush_enable'])),
+				['config.update', ['webpush_enable', $this->config['wpn_webpush_enable']]],
+			]],
+			['if', [
+				(isset($this->config['webpush_vapid_public'])),
```
Comment (iMattPro) [inline migrations/handle_subscriptions.php:35]: For these keys, should we be concerned about overwriting, like if the user somehow already added keys to the core's web push? Should we only replace them with the keys from the extension if the core's key is empty?

File: migrations/handle_subscriptions.php [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+				['config.update', ['webpush_vapid_public', $this->config['wpn_webpush_vapid_public']]],
+			]],
+			['if', [
+				(isset($this->config['webpush_vapid_private'])),
+				['config.update', ['webpush_vapid_private', $this->config['wpn_webpush_vapid_private']]],
+			]],
+		];
+	}
+
+	public function update_subscriptions()
```
Comment (iMattPro) [inline migrations/handle_subscriptions.php:45]: Lets just add some docblock comments to this and the function below so we can remember what they are for/doing in the future.

**Pull request merged**